app/trainer.py

```diff
@@ -0,0 +1,329 @@
+                                  index=False,
+                                  header='text')
+
+        # dev_values = test_values.sample(frac=0.50, random_state=42)
+        # test_values = test_values.drop(dev_values.index)
+
+        train_examples = create_examples(train_values, num_classes, 'train')
+        # eval_examples = create_examples(dev_values, num_classes, 'dev')
+
+        params = dict(num_classes=num_classes,
+                      learning_rate=LEARNING_RATE,
+                      num_warmup_steps=NUM_WARMUP_STEPS,
+                      dropout=DROPOUT,
+                      class_based_attention=True,
+                      shared_size=SHARED_SIZE,
+                      num_train_steps=num_train_steps)
+
+        run_config = tf.estimator.RunConfig(model_dir=train_path,
+                                            save_checkpoints_steps=min(
+                                                [num_train_steps, 500]))
+
+        model_fn = class_based_attention.model_fn_builder(use_tpu=False)
+
+        estimator = tf.estimator.Estimator(model_fn=model_fn,
+                                           config=run_config,
+                                           params=params)
+
+        status_logger('***** Started training at {} *****'.format(
+            datetime.now()))
+        train_input_fn = input_fn_builder(train_examples,
+                                          num_classes,
+                                          embedder,
+                                          is_training=True,
+                                          drop_remainder=False)
+
+        saved_model_path = os.path.join(train_path, 'saved_models')
+        estimator.train(input_fn=train_input_fn, steps=num_train_steps)
+        save_model(saved_model_path, estimator)
+        status_logger('***** Finished training at {} *****'.format(
+            datetime.now()))
+
+        copy_tree(
+            os.path.join(saved_model_path,
+                         sorted(os.listdir(saved_model_path), reverse=True)[0]),
+            instance_path)
+
+        c = Classifier(self.base_classifier_dir,
+                       self.model_name,
+                       forced_instance=timestamp)
+        c.refresh_thresholds([seqs[i] for i in test_values.index],
+                             [training_labels[i] for i in test_values.index])
+
+        config['is_released'] = True
+        with open(os.path.join(instance_path, 'config.json'), 'w') as f:
+            json.dump(config, f)
+        ClassifierCache.clear(self.base_classifier_dir, self.model_name)
+        return c
+
+
+DEFAULT_BERT = 'https://tfhub.dev/google/bert_uncased_L-12_H-768_A-12/1'
+
+
+class _TrainModel(tasks.TaskProvider):
+
+    def __init__(self, json: Any):
+        super().__init__(json)
+        self.base_classifier_dir = json['base_classifier_dir']
+        self.model = json['model']
+        self.vocab = json['vocab'] if 'vocab' in json else None
```

Comment:
Vocab is always a property of the bert base model. It shouldn't be possible to load a vocab file with a BERT model to which it doesn't belong

Reply:
vocab is about the model output (e.g. 'Torture'), not the BERT vocab. I added a comment.

Reply:
let's please change vocab to labels. In ML these two have very different meaning

Reply:
Changed everywhere